core/silkworm/consensus/engine.hpp

```diff
@@ -0,0 +1,94 @@
+/*
+   Copyright 2021 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#ifndef SILKWORM_CONSENSUS_ENGINE_HPP_
+#define SILKWORM_CONSENSUS_ENGINE_HPP_
+
+#include <unordered_map>
+#include <vector>
+
+#include <silkworm/consensus/validation.hpp>
+#include <silkworm/execution/state_pool.hpp>
+#include <silkworm/state/intra_block_state.hpp>
+#include <silkworm/state/state.hpp>
+#include <silkworm/types/receipt.hpp>
+
+namespace silkworm::consensus {
+
+class IConsensusEngine {
+  public:
+    explicit IConsensusEngine(const ChainConfig& chain_config) : chain_config_{chain_config} {};
+    virtual ~IConsensusEngine() = default;
+
+    //! \brief Performs validation of block header & body that can be done prior to sender recovery and execution.
+    //! \brief See [YP] Sections 4.3.2 "Holistic Validity", 4.3.4 "Block Header Validity", and 11.1 "Ommer Validation".
+    //! \param [in] block: block to pre-validate.
+    //! \param [in] state: current state.
+    //! \note Shouldn't be used for genesis block.
+    virtual ValidationResult validate_block(const Block& block, State& state) = 0;
```

Comment:
I'd keep `pre_validate_block` because this is not the full validation. There are some other checks performed in `ExecutionProcessor::validate_transaction` & `execute_and_write_block`.